File: leetcode/Algorithms/386.LexicographicalNumbers/Solution.py
class Solution(object):
    def lexicalOrder(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        result = []
        current = 1
        for _ in xrange(n):
            result.append(current)
            if current * 10 <= n:
                current *= 10
            elif current % 10 != 9 and current + 1 <= n:
                current += 1
            else:
                current /= 10
                while current % 10 == 9:
                    current /= 10
                current += 1
        return result

File: leetcode/Algorithms/386.LexicographicalNumbers/Solution.java
public class Solution {
    private List<Integer> result;
    private int n;

    public List<Integer> lexicalOrder(int n) {
        this.result = new ArrayList<>();
        this.n = n;
        nextRound(1);

        return result;
    }

    private void nextRound(int current) {
        result.add(current);
        if (current * 10 <= n) nextRound(10 * current);
        if (current < n && current % 10 < 9) nextRound(current + 1);
    }
}
